File: blind75/two_pointers/3sum.py
# (1) if left-most value is positive, all other numbers will
#     be positive because nums is sorted, so no other combinations
#     will sum to 0 from this point on - we can therefore short-circuit
# (2) avoid duplicates by incrementing l & m pointers if previous num 
#     was the same num; note that m pointer increment happens at the end
#     of the loop so that the m < r check applies first before the new
#     m value is used (if you do it within the while-body, m can equal r 
#     and you use an invalid triplet)
# note: there is probably a way to use binary search instead of two-pointer
#       which may be faster

class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        nums = sorted(nums)
        result = []
           
        for l in range(len(nums)):
            if nums[l] > 0: # (1)
                break
                
            if l != 0 and nums[l] == nums[l-1]: # (2)
                continue
                
            r = len(nums) - 1
            m = l + 1
            
            while m < r:
                total = nums[l] + nums[m] + nums[r]
                
                if total > 0:
                    r -= 1
                    
                elif total < 0:
                    m += 1
                    
                else: #elif total == 0:
                    result.append([nums[l],nums[m],nums[r]])
                    m += 1
                    r -= 1
                    while m < r and nums[m] == nums[m-1]: # (2)
                        m += 1
                        
        return result      

File: grind75/review/3sum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();

        for (int first = 0; first < nums.length; first++) {

            if (first > 0 && nums[first] == nums[first - 1]) {
                continue;
            }

            int second = first + 1;
            int third = nums.length - 1;

            while (second < third) {

                if (third < nums.length - 1 && nums[third] == nums[third + 1]) {
                    third--;
                    continue;
                }

                int sum = nums[first] + nums[second] + nums[third];

                if (sum == 0) {
                    List<Integer> triplet = new ArrayList<>();
                    triplet.add(nums[first]);
                    triplet.add(nums[second]);
                    triplet.add(nums[third]);
                    res.add(triplet);
                    second++;
                    third--;
                }

                else if (sum < 0) {
                    second++;
                }

                else {
                    third--;
                }
            }

        }

        return res;

    }
}